src/server/components/charts-engine/runners/control.ts

```diff
@@ -0,0 +1,226 @@
+            stype: ControlType.Dash,
+        },
+        publicAuthor: config.publicAuthor,
+    } as ResolvedConfig;
+
+    const hrStart = process.hrtime();
+
+    const {params, actionParams, widgetConfig} = req.body;
+
+    const iamToken = res?.locals?.iamToken ?? req.headers[ctx.config.headersMap.subjectToken];
+
+    const controlBuilder = await getControlBuilder({
+        config: generatedConfig,
+    });
+    const processorParams: Omit<ProcessorParams, 'ctx'> = {
+        chartsEngine,
+        paramsOverride: params,
+        actionParamsOverride: actionParams,
+        widgetConfig,
+        userLang: res.locals && res.locals.lang,
+        userLogin: res.locals && res.locals.login,
+        userId: res.locals && res.locals.userId,
+        subrequestHeaders: res.locals.subrequestHeaders,
+        req,
+        iamToken,
+        isEditMode: Boolean(res.locals.editMode),
+        configResolving,
+        cacheToken: req.headers['x-charts-cache-token'] || null,
+        builder: controlBuilder,
+    };
+
+    if (generatedConfig) {
+        processorParams.configOverride = generatedConfig;
+    }
+
+    const configWorkbook = workbookId ?? config.workbookId;
+    if (configWorkbook) {
+        processorParams.workbookId = configWorkbook;
+    }
+
+    if (req.body.uiOnly) {
+        processorParams.uiOnly = true;
+    }
+
+    processorParams.responseOptions = req.body.responseOptions || {};
+
+    if (
+        processorParams.responseOptions &&
+        typeof processorParams.responseOptions.includeLogs === 'undefined'
+    ) {
+        processorParams.responseOptions.includeLogs = true;
+    }
+
+    ctx.log('ControlRunner::PreRun', {duration: getDuration(hrStart)});
+
+    const showChartsEngineDebugInfo = Boolean(
+        isEnabledServerFeature(ctx, Feature.ShowChartsEngineDebugInfo),
+    );
+
+    return ctx
+        .call('engineProcessing', (cx) => {
+            return Processor.process({...processorParams, ctx: cx})
+                .then((result) => {
+                    cx.log('ControlRunner::FullRun', {duration: getDuration(hrStart)});
+
+                    res.setHeader('chart-runner-type', 'controls');
+
+                    if (result) {
+                        if (
+                            'logs_v2' in result &&
+                            (!res.locals.editMode || !showChartsEngineDebugInfo)
+                        ) {
+                            delete result.logs_v2;
+                        }
+
+                        if ('error' in result) {
+                            const resultCopy = {...result};
+
+                            if ('_confStorageConfig' in resultCopy) {
+                                delete resultCopy._confStorageConfig;
+                            }
+
+                            const logError = prepareErrorForLogger(result.error);
+
+                            cx.log('PROCESSED_WITH_ERRORS', {error: logError});
+
+                            let statusCode = 500;
```

Comment:
This one (and  below) code is the same for all runners. It would be nice to extract it into an external function and reuse it.

Because it's to complicate for support as copy. There are so many places for bugs.